File: ocaml/003/003.ml
let prime_factors n =
  let rec pf c p =
    match c with
    | c when c < (p * p) -> [c]
    | c when (c mod p) = 0 -> [p] @ (pf (c / p) p)
    | _ -> pf c (p + 1)
  in
  pf n 2;;

let factors = prime_factors 600851475143;;

Printf.printf "%d\n" (List.fold_left max 0 factors)

File: scala/003/003.scala
object Euler {
  def main(args: Array[String]) {
    println(primeFactors(600851475143L).max);
  }

  def primeFactors(n: Long): List[Long] = {
    def pf(tuple: (Long, Long)): List[Long] = {
      tuple match {
        case (n, p) if (n < p * p)  => List[Long](n)
        case (n, p) if (n % p == 0) => List[Long](p) ++ pf((n / p, p))
        case (n, p)                 => pf((n, p + 1))
      }
    }
    pf((n, 2))
  }
}
